### PR TITLE
[ graph ] Divide addLossLayer process when compiling inference model graph @open sesame 01/15 14:27

### PR DESCRIPTION
**Pull request by skykongkong8**
- Inference model does not require label and loss layer.
- Therefore, we can just simply ignore addLossLayer during the inference mode.
- Add unittest to verify such.

**Self evaluation:**
1. Build test:     [X]Passed [ ]Failed [ ]Skipped
2. Run test:     [X]Passed [ ]Failed [ ]Skipped

FYI) Following is the output of : 
```C++
TEST(nntrainerGraphUnitTest, NoLossLayerWhenInferenceMode) {
...
   model->summarize(std::cout, ML_TRAIN_SUMMARY_MODEL);
...
}
```
```bash
================================================================================
          Layer name          Layer type    Output dimension         Input layer
================================================================================
              input0               input           1:1:1:256                    
--------------------------------------------------------------------------------
    fully_connected0     fully_connected          1:1:1:1024              input0
--------------------------------------------------------------------------------
    fully_connected1     fully_connected          1:1:1:1024    fully_connected0
--------------------------------------------------------------------------------
    fully_connected2     fully_connected          1:1:1:1024    fully_connected1
--------------------------------------------------------------------------------
    fully_connected3     fully_connected           1:1:1:100    fully_connected2
================================================================================
```

### CODE REVIEW

File: test/unittest/unittest_nntrainer_graph.cpp [hunk 1]
```diff
@@ -368,6 +368,55 @@ TEST(nntrainerGraphUnitTest, call_functions) {
...
+    "fully_connected",
+    {withKey("unit", 100), withKey("weight_initializer", "xavier_uniform"),
+     withKey("bias_initializer", "zeros")}));
+
+  model->setProperty({withKey("batch_size", 1), withKey("epochs", 1),
+                      withKey("memory_swap", "false"),
+                      withKey("model_tensor_type", "FP32-FP32")});
+
+  auto optimizer = ml::train::createOptimizer("sgd", {"learning_rate=0.001"});
+  model->setOptimizer(std::move(optimizer));
```
Comment (EunjuYang) [inline test/unittest/unittest_nntrainer_graph.cpp:394]: I think the optimizer is unnecessary as well for the `INFERENCE` mode.

**Review by EunjuYang — APPROVED** [review]
Nice work! 👍

File: test/unittest/unittest_nntrainer_graph.cpp (hunk 1 above)
Comment (skykongkong8) [reply test/unittest/unittest_nntrainer_graph.cpp:394]: You have a point! Let me fix this one.

File: test/unittest/unittest_nntrainer_graph.cpp [hunk 2]
```diff
@@ -368,6 +368,52 @@ TEST(nntrainerGraphUnitTest, call_functions) {
...
+                      withKey("model_tensor_type", "FP32-FP32")});
+
+  int status = model->compile(ml::train::ExecutionMode::INFERENCE);
+  EXPECT_EQ(status, ML_ERROR_NONE);
+
+  status = model->initialize(ml::train::ExecutionMode::INFERENCE);
+  EXPECT_EQ(status, ML_ERROR_NONE);
+
+  float input[256];
+  float label[1];
```
Comment (DonghakPark) [inline test/unittest/unittest_nntrainer_graph.cpp:400]: is this label need ?

**Review by DonghakPark — APPROVED** [review]
Good Point!! LGTM!

File: test/unittest/unittest_nntrainer_graph.cpp (hunk 2 above)
Comment (skykongkong8) [reply test/unittest/unittest_nntrainer_graph.cpp]: I missed it😂 Thanks!

**Review by EunjuYang — APPROVED** [review]

**Comment by skykongkong8** [comment]
It seems issue raised from #2867 is occurring here as well

**Comment by djeong20** [comment]
> It seems issue raised from #2867 is occurring here as well

please re-run the action to pass as for now.

**Review by lhs8928 — APPROVED** [review]
LGTM

**Comment by skykongkong8** [comment]
For the static checker error, it seems there's an similar problem that @djeong20 raised from  #2870 is occuring for the PR as well.
With the Static checkers and verifier's suggestion, I could not avoid C++ Format Checker, and with C++ Format Checker's suggesion, I could not avoid Static checkers and verifier error.

File: nntrainer/graph/network_graph.cpp [hunk 3]
```diff
@@ -65,14 +65,21 @@ int NetworkGraph::compile(const std::string &loss_type) {
...
+      /// @todo realize loss beforehand
+      status = addLossLayer(loss_type);
+      NN_RETURN_STATUS();
+    } catch (const std::exception &e) {
+      ml_loge("%s", e.what());
+      status = ML_ERROR_INVALID_PARAMETER;
+      NN_RETURN_STATUS();
+    }
+  } else {
+    if (!loss_type.empty()) {
```
Comment (baek2sm) [inline nntrainer/graph/network_graph.cpp:79]: Wouldn't it be better to allow it to run even with loss layer in inference mode for evaluation?
I think someone can add a loss layer in inference mode to evaluate the performance of the model without training.

Comment (skykongkong8) [reply nntrainer/graph/network_graph.cpp:79]: Well, you have a point. Maybe we should discuss with @jijoongmoon about this point.
